govcd/nsxt_l2_vpn_tunnel.go

```diff
@@ -0,0 +1,282 @@
+	if egw.EdgeGateway == nil || egw.client == nil || egw.EdgeGateway.ID == "" {
+		return nil, fmt.Errorf("cannot get L2 VPN tunnel for NSX-T Edge Gateway without ID")
+	}
+
+	client := egw.client
+	endpoint := types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointEdgeGatewayL2VpnTunnel
+	apiVersion, err := client.getOpenApiHighestElevatedVersion(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	urlRef, err := client.OpenApiBuildEndpoint(fmt.Sprintf(endpoint, egw.EdgeGateway.ID), id)
+	if err != nil {
+		return nil, err
+	}
+
+	tunnel := &NsxtL2VpnTunnel{
+		client:        egw.client,
+		edgeGatewayId: egw.EdgeGateway.ID,
+	}
+	err = client.OpenApiGetItem(apiVersion, urlRef, nil, &tunnel.NsxtL2VpnTunnel, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return tunnel, nil
+}
+
+// Statistics retrieves connection statistics for a given L2 VPN Tunnel configured on an Edge Gateway.
+func (l2Vpn *NsxtL2VpnTunnel) Statistics() (*types.EdgeL2VpnTunnelStatistics, error) {
+	client := l2Vpn.client
+	endpoint := types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointEdgeGatewayL2VpnTunnelStatistics
+	apiVersion, err := client.getOpenApiHighestElevatedVersion(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	urlRef, err := client.OpenApiBuildEndpoint(fmt.Sprintf(endpoint, l2Vpn.edgeGatewayId, l2Vpn.NsxtL2VpnTunnel.ID))
+	if err != nil {
+		return nil, err
+	}
+
+	statistics := &types.EdgeL2VpnTunnelStatistics{}
+	err = client.OpenApiGetItem(apiVersion, urlRef, nil, &statistics, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return statistics, nil
+}
+
+// Status retrieves status of a given L2 VPN Tunnel.
+func (l2Vpn *NsxtL2VpnTunnel) Status() (*types.EdgeL2VpnTunnelStatus, error) {
+	client := l2Vpn.client
+	endpoint := types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointEdgeGatewayL2VpnTunnelStatus
+	apiVersion, err := client.getOpenApiHighestElevatedVersion(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	urlRef, err := client.OpenApiBuildEndpoint(fmt.Sprintf(endpoint, l2Vpn.edgeGatewayId, l2Vpn.NsxtL2VpnTunnel.ID))
+	if err != nil {
+		return nil, err
+	}
+
+	status := &types.EdgeL2VpnTunnelStatus{}
+	err = client.OpenApiGetItem(apiVersion, urlRef, nil, &status, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return status, nil
+}
+
+// Update updates the L2 VPN tunnel with the provided parameters as the argument
+func (l2Vpn *NsxtL2VpnTunnel) Update(tunnelParams *types.NsxtL2VpnTunnel) (*NsxtL2VpnTunnel, error) {
+	if l2Vpn.NsxtL2VpnTunnel.SessionMode != tunnelParams.SessionMode {
+		return nil, fmt.Errorf("error updating the L2 VPN Tunnel: session mode can't be changed after creation")
+	}
+
+	if tunnelParams.SessionMode == "CLIENT" && !tunnelParams.Enabled {
+		// There is a known bug up to 10.5.0, the CLIENT sessions can't be
+		// disabled and can result in unexpected behaviour for the following
+		// operations
+		if l2Vpn.client.APIVCDMaxVersionIs("<= 38.0") {
+			return nil, fmt.Errorf("client sessions can't be disabled")
```

Comment:
What about extending the error message to describe why (prior to which version) it can't be disabled?

Reply:
I think I'll leave it, it's not certain if this will be allowed in future versions

Reply:
Done, but I'm not sure if they add the option in future versions as it might be intended